### DISCUSSION TITLE
What is the pretaproot module for

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-miniscript#385] [Q&A] [answered]
The `pretaproot` module, and specifically the `PreTaprootDescriptor` type, contains duplicate code from the `Descrpitor` type. It is not immediately obvious what this module is for?

Would adding this method to `Descriptor` enable users of the library to check a descriptor before using it if they specifically needed a pre-taproot descriptor?
```rust
    /// Returns true if this desciptor represents scripts from _before_ Taproot was enabled.
    pub fn is_pre_taproot(&self) -> bool {
        use Descriptor::*;

        match self {
            Bare(_) | Pkh(_) | Wpkh(_) | Sh(_) | Wsh(_) => true,
            Tr(_) => false,
        }
    }
```

**Comment by apoelstra** [comment, marked as answer]
@tcharding there are several different methods in `PreTaprootDescriptor` vs `Descriptor`. In particular things that did not return a ``Result` pre-Taproot now do.

**Reply by tcharding** [reply]
oooo, I missed that. Thanks.
